flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,275 @@
+    public static final int MIN_TOTAL_BYTES = 32 * 1024 * 1024;
+
+    /**
+     * Memory size in bytes can be allocated from this buffer pool for a single request (8M is for
+     * better sequential read).
+     */
+    public static final int NUM_BYTES_PER_REQUEST = 8 * 1024 * 1024;
+
+    /** Total direct memory size in bytes can can be allocated and used by this buffer pool. */
+    private final long totalBytes;
+
+    /**
+     * Maximum time to wait in milliseconds when requesting read buffers from this buffer pool
+     * before throwing an exception.
+     */
+    private final long requestTimeout;
+
+    /** The number of total buffers in this buffer pool. */
+    private final int numTotalBuffers;
+
+    /** Size of each buffer in bytes in this buffer pool. */
+    private final int bufferSize;
+
+    /** The number of buffers to be returned for a single request. */
+    private final int numBuffersPerRequest;
+
+    /** All available buffers in this buffer pool currently. */
+    @GuardedBy("buffers")
+    private final Queue<MemorySegment> buffers = new ArrayDeque<>();
+
+    /** Whether this buffer pool has been destroyed or not. */
+    @GuardedBy("buffers")
+    private boolean destroyed;
+
+    /** Whether this buffer pool has been initialized or not. */
+    @GuardedBy("buffers")
+    private boolean initialized;
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize) {
+        // 5 min default buffer request timeout
+        this(totalBytes, bufferSize, 5 * 60 * 1000);
+    }
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
+        checkArgument(totalBytes > 0, "Total memory size must be positive.");
+        checkArgument(bufferSize > 0, "Size of buffer must be positive.");
+        checkArgument(requestTimeout > 0, "Request timeout must be positive.");
+
+        this.totalBytes = totalBytes;
+        this.bufferSize = bufferSize;
+        this.requestTimeout = requestTimeout;
+
+        this.numTotalBuffers = (int) Math.min(totalBytes / bufferSize, Integer.MAX_VALUE);
+        this.numBuffersPerRequest =
+                Math.min(numTotalBuffers, Math.max(1, NUM_BYTES_PER_REQUEST / bufferSize));
+    }
+
+    @VisibleForTesting
+    public long getTotalBytes() {
+        return totalBytes;
+    }
+
+    @VisibleForTesting
+    public int getNumBuffersPerRequest() {
+        return numBuffersPerRequest;
+    }
+
+    @VisibleForTesting
+    public int getNumTotalBuffers() {
+        return numTotalBuffers;
+    }
+
+    @VisibleForTesting
+    public int getAvailableBuffers() {
+        synchronized (buffers) {
+            return buffers.size();
+        }
+    }
+
+    public int getMaxConcurrentRequests() {
+        return numBuffersPerRequest > 0 ? numTotalBuffers / numBuffersPerRequest : 0;
+    }
+
+    /** Initializes this buffer pool which allocates all the buffers. */
+    public void initialize() {
```

Comment:
Could we make it private and `testInitializeAfterDestroy` can be replaced with a `testRequestBuffersAfterDestroy`?
For those case that need the pool to be initialized, we can use `execute` to trigger the initialization.